src/test/java/com/snowflake/kafka/connector/internal/StageFilesProcessorTest.java

```diff
@@ -0,0 +1,411 @@
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiConsumer;
+import net.snowflake.client.jdbc.internal.joda.time.DateTime;
+import net.snowflake.client.jdbc.internal.joda.time.DateTimeZone;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Mockito;
+
+class StageFilesProcessorTest {
+
+  private static final String PIPE_NAME = "testPipe";
+  private static final String TABLE_NAME = "testTable";
+  private static final String STAGE_NAME = "testStage";
+  private static final String PREFIX = "prefix";
+
+  private SnowflakeConnectionService conn;
+  private SnowflakeIngestionService ingestionService;
+  private AtomicLong currentTime;
+  private StageFilesProcessor victim;
+  private StageFilesProcessor.ProgressRegisterImpl register;
+  private SnowflakeTelemetryPipeCreation pipeCreation;
+  private SnowflakeTelemetryPipeStatus pipeTelemetry;
+  private SnowflakeTelemetryService telemetryService;
+
+  private PipeProgressRegistryTelemetry telemetry;
+  private AtomicInteger loops;
+  private AtomicReference<BiConsumer<Integer, Long>> nextTickCallback;
+  private AtomicReference<ScheduledFuture<?>> scheduledFuture;
+
+  @BeforeEach
+  void setUp() {
+    loops = new AtomicInteger(0);
+    currentTime = new AtomicLong(new DateTime(2000, 1, 10, 12, 0, DateTimeZone.UTC).getMillis());
+    nextTickCallback = new AtomicReference<>();
+    scheduledFuture = new AtomicReference<>();
+
+    conn = Mockito.mock(SnowflakeConnectionService.class);
+    ingestionService = Mockito.mock(SnowflakeIngestionService.class);
+    pipeCreation = new SnowflakeTelemetryPipeCreation(TABLE_NAME, STAGE_NAME, PIPE_NAME);
+    pipeTelemetry = Mockito.mock(SnowflakeTelemetryPipeStatus.class);
+    telemetryService = Mockito.mock(SnowflakeTelemetryService.class);
+    telemetry = new PipeProgressRegistryTelemetry(pipeCreation, pipeTelemetry, telemetryService);
+
+    victim =
+        new StageFilesProcessor(
+            PIPE_NAME,
+            TABLE_NAME,
+            STAGE_NAME,
+            PREFIX,
+            conn,
+            ingestionService,
+            pipeTelemetry,
+            telemetryService,
+            createTestScheduler(loops, currentTime, nextTickCallback, scheduledFuture),
+            currentTime::get);
+    register = new StageFilesProcessor.ProgressRegisterImpl(victim);
+  }
+
+  @Test
+  void filesProcessWillTerminateOnStopSignal() {
+    StageFilesProcessor.ProgressRegister client = victim.trackFilesAsync();
+    client.close();
+    verify(scheduledFuture.get(), times(1)).cancel(true);
+  }
+
+  @Test
+  void fileProcessorWillTryToGetInitialStateFromStageExactlyOnce() {
+    // lets simulate 1 hour (60 cleanup cycles)
+    loops.set(60);
```

Comment:
Nit: it looks a little bit complex to grasp.
First we create a scheduler with `loops` passed as an argument. Then we alter the `loops` values so that during `victim.trackFiles` it's taken into account since we propagated it earlier.
Will it be simpler to have a method, let's say `createStageFilesProcessor`, that takes number of iterations as an argument? So we create it in each test without any indirection involving atomics?